### ISSUE TITLE
Single Image Inference

### ISSUE THREAD

**Issue by GreatGameDota** [Zengyi-Qin/MonoGRNet#37]
How do I use the pretrained model for single image inference? With no validation data.  
Data directory format:
```
data
  training
    image_2
      test_image.jpg
```
I also have ```eval_iter``` set to 1 in ```hypes/kittiBox.json```  
I have tried doing the inference on a cpu by running ```python train.py``` but errors out due to memory. And I tried running the same thing on a gpu but it just runs indefinitely with no error messages.

**Comment by Zengyi-Qin** [comment]
Hi, thanks for your interest. The `data/KittiBox/val.txt` points to the testing images. For example, in its first line the `training/image_2/000001.png training/label_2/000001.txt` indicates `data/KittiBox/training/image_2/000001.png` will be used for testing. You can change the image name to your own image.  You can keep `training/label_2/000001.txt` unchanged since the labels are not required in inference.

The result (in txt) will be stored in `runs`

**Comment by GreatGameDota** [comment]
Thanks for the quick reply! This is my current file format:
<img width="174" alt="72e85691b9dd4e8d198753ac44c49751" src="https://user-images.githubusercontent.com/17858650/69496425-b3589900-0e97-11ea-94e3-59d4f9850fcd.png">
And this is what I have in my val.txt: ```training/image_2/ID_00a798026.jpg training/label_2/ID_00a798026.txt```
I then run ```python setup.py``` and ```python train.py```
And get this error:  
```File "/content/hypes/../inputs/kitti_input.py", line 374, in start_enqueuing_threads```
```    data = gen.next()```
```AttributeError: 'generator' object has no attribute 'next'```

**Comment by Zengyi-Qin** [comment]
The `train.py` will start with evaluation ( evaluates at step 0, step eval_iter, step eval_iter * 2 , and so forth). A single image will take <0.1s, but the initializaiton of the pipeline would take some time. But I believe 3 min is enough. As for the `next` attribute, I've encountered it before. It's due to a wrong version of Python (Python2.7 is required). I remember Python3.6 would get in to trouble cuz the data generator (a tool that iteratively feeds data to the pipeline) does not have the `next` attribute.

**Comment by GreatGameDota** [comment]
I just found the solution to that right before reading this. I just changed ```data = gen.next()``` to ```data = next(gen)``` and now I'm getting an error saying there is no calib txt file for the test image. ```AssertionError: File does not exist: /content/data/KittiBox/training/calib/ID_00a798026.txt```
EDIT: I was able to get past the error by adding the txt file and filling it with:
```
P0: 7.070493000000e+02 0.000000000000e+00 6.040814000000e+02 0.000000000000e+00 0.000000000000e+00 7.070493000000e+02 1.805066000000e+02 0.000000000000e+00 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00 0.000000000000e+00
P1: 7.070493000000e+02 0.000000000000e+00 6.040814000000e+02 -3.797842000000e+02 0.000000000000e+00 7.070493000000e+02 1.805066000000e+02 0.000000000000e+00 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00 0.000000000000e+00
P2: 7.070493000000e+02 0.000000000000e+00 6.040814000000e+02 4.575831000000e+01 0.000000000000e+00 7.070493000000e+02 1.805066000000e+02 -3.454157000000e-01 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00 4.981016000000e-03
P3: 7.070493000000e+02 0.000000000000e+00 6.040814000000e+02 -3.341081000000e+02 0.000000000000e+00 7.070493000000e+02 1.805066000000e+02 2.330660000000e+00 0.000000000000e+00 0.000000000000e+00 1.000000000000e+00 3.201153000000e-03
R0_rect: 9.999128000000e-01 1.009263000000e-02 -8.511932000000e-03 -1.012729000000e-02 9.999406000000e-01 -4.037671000000e-03 8.470675000000e-03 4.123522000000e-03 9.999556000000e-01
Tr_velo_to_cam: 6.927964000000e-03 -9.999722000000e-01 -2.757829000000e-03 -2.457729000000e-02 -1.162982000000e-03 2.749836000000e-03 -9.999955000000e-01 -6.127237000000e-02 9.999753000000e-01 6.931141000000e-03 -1.143899000000e-03 -3.321029000000e-01
Tr_imu_to_velo: 9.999976000000e-01 7.553071000000e-04 -2.035826000000e-03 -8.086759000000e-01 -7.854027000000e-04 9.998898000000e-01 -1.482298000000e-02 3.195559000000e-01 2.024406000000e-03 1.482454000000e-02 9.998881000000e-01 -7.997231000000e-01
```
And this runs with no other errors but then I get a lot of warnings saying that the program failed to allocate memory and it eventually ends.

**Comment by Zengyi-Qin** [comment]
yes, there should be a calib file.  I would suggest you to resize (or crop out a preferred region and resize) the image to height 360 * width 1200 before putting it to the folder, because the calib file is for images approximately of this size. 

I think you're very close to the destination as no error occured. Regarding the memory, are you using cpus? I didn't try inference on cpus. The pipeline works on a single gpu (~10G memory)

**Comment by Zengyi-Qin** [comment]
Another way that might solve the memory issue: if training is not required, I would suggest to comment out some lines in the `run_training` function in `include/tensorvision/train.py`. Comment the lines from `if step % display_iter:` to ` _write_eval_dict_to_summary(eval_dict, 'Eval/smooth', summary_writer, step)`. The remaining part of the `run_training` funciton is thus only for evaluation, and does not require as much memory as the training does

**Comment by GreatGameDota** [comment]
I am using a GPU with 12gb free ram. I resized the image from 3384x1341 to 1200x476 so it might work now. If that doesn't work I'll comment out those lines. Thanks!  
EDIT: So I commented out the unneeded  lines and fixed the new errors except for one: It is saying the ```include/utils/stitch_wrapper``` wasn't compiled by ```submodules/KittiEvaluation``` and to ```cd submodules/KittiEvaluation && make``` but it should do that when I run ```setup.py```
